packages/react-dom/src/client/ReactInputSelection.js

```diff
@@ -49,34 +137,38 @@ export function getSelectionInformation() {
  * nodes and place them back in, resulting in focus being lost.
  */
 export function restoreSelection(priorSelectionInformation) {
-  const curFocusedElem = getActiveElement();
-  const priorFocusedElem = priorSelectionInformation.focusedElem;
-  const priorSelectionRange = priorSelectionInformation.selectionRange;
-  if (curFocusedElem !== priorFocusedElem && isInDocument(priorFocusedElem)) {
-    if (hasSelectionCapabilities(priorFocusedElem)) {
-      setSelection(priorFocusedElem, priorSelectionRange);
-    }
-
-    // Focusing a node can change the scroll position, which is undesirable
-    const ancestors = [];
-    let ancestor = priorFocusedElem;
-    while ((ancestor = ancestor.parentNode)) {
-      if (ancestor.nodeType === ELEMENT_NODE) {
-        ancestors.push({
-          element: ancestor,
-          left: ancestor.scrollLeft,
-          top: ancestor.scrollTop,
-        });
+  const priorActiveElement = priorSelectionInformation.activeElement;
+  const elementSelections = priorSelectionInformation.elementSelections;
+  let curActiveElement = getActiveElementDeep();
+  const isActiveElementOnlySelection =
+    elementSelections.length === 1 &&
+    elementSelections[0] === priorActiveElement;
+  if (
+    !isInDocument(priorActiveElement) ||
+    priorActiveElement === priorActiveElement.ownerDocument.body ||
+    (isActiveElementOnlySelection && curActiveElement === priorActiveElement)
+  ) {
+    return;
+  }
+  elementSelections.forEach(function(selection) {
+    const element = selection.element;
+    if (
+      isInDocument(element) &&
+      getActiveElement(element.ownerDocument) !== element
+    ) {
+      setSelection(element, selection.selectionRange);
+      if (element !== priorActiveElement) {
+        focusNodePreservingScroll(element);
+        curActiveElement = element;
       }
     }
+  });
 
-    focusNode(priorFocusedElem);
-
-    for (let i = 0; i < ancestors.length; i++) {
-      const info = ancestors[i];
-      info.element.scrollLeft = info.left;
-      info.element.scrollTop = info.top;
-    }
+  if (
+    curActiveElement !== priorActiveElement &&
+    isInDocument(priorActiveElement)
```

Comment:
Possible to recycle this value from the check on line 147?

Reply:
The check is actually useless, because if `priorActiveElement` is not in document, the function will early return and never reach this if block.